### ISSUE TITLE
numpy.ndarray.newbyteorder much slower in numpy1.10

### ISSUE THREAD

**Issue by cpaulik** [numpy/numpy#6658]
Recently @christophreimer noticed that one of our dataset readers got noticeably slower in a new virtual environment.
This could be tracked down to calls of `numpy.ndarray.newbyteorder`. 
I'm running the anaconda builds for numpy 1.10.1 and numpy 1.9.3 for comparison:

### numpy 1.9.3
```python
Python 2.7.10 |Continuum Analytics, Inc.| (default, Oct 19 2015, 18:04:42) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import numpy as np

In [2]: a = np.arange(1000)

In [3]: %timeit b = a.newbyteorder('L')
The slowest run took 45.18 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 554 ns per loop

In [4]: %timeit b = a.newbyteorder('B')
The slowest run took 22.81 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 564 ns per loop

In [5]: np.__version__
Out[5]: '1.9.3'
```

### numpy 1.10.1
```python
Python 2.7.10 |Continuum Analytics, Inc.| (default, Oct 19 2015, 18:04:42) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import numpy as np

In [2]: np.__version__
Out[2]: '1.10.1'

In [3]: a = np.arange(1000)

In [4]: %timeit b = a.newbyteorder('B')
The slowest run took 9.73 times longer than the fastest. This could mean that an intermediate result is being cached 
100000 loops, best of 3: 11.1 µs per loop

In [5]: %timeit b = a.newbyteorder('L')
The slowest run took 6.45 times longer than the fastest. This could mean that an intermediate result is being cached 
100000 loops, best of 3: 11.2 µs per loop
```

**Comment by ahaldane** [comment]
I can confirm there is a slowdown in 1.10.1,  but it appears to be fixed in 1.10.2 (to be relased soon). 

It's possibly related to #6467 even though there are no structured arrays here - a number of code paths which called python code have been eliminated in 1.10.2.

**Comment by seberg** [comment]
Lets close this then. Please reopen/comment if the 1.10.2 release does not fix it for you. Thanks for the report!
